### PR TITLE
bugfix: view: Fix Zulip crash during search.

### LINKED ISSUE #923: crash while using search
I was using the zulip search box [q] and I went to edit a former stream search. I deleted the entire search and pressed enter. Then the app crashed. I was unable to replicate this incident so I'm not sure if I should be creating an issue for it?

I am running Zulip Terminal 0.6.0+git on mac in a Z shell....The following was in my zulip-terminal-tracebacks.log:

{e}

'<' not supported between instances of 'NoneType' and 'int'
Traceback (most recent call last):
  File "/Users/kellyfoulk/Documents/code/zulip/zulip-terminal/zulipterminal/cli/run.py", line 434, in main
    Controller(zuliprc_path,
  File "/Users/kellyfoulk/Documents/code/zulip/zulip-terminal/zulipterminal/core.py", line 438, in main
    self.loop.run()
  File "/Users/kellyfoulk/Documents/code/zulip/zulip-terminal/zt_venv/lib/python3.9/site-packages/urwid/main_loop.py", line 287, in run
    self._run()
  File "/Users/kellyfoulk/Documents/code/zulip/zulip-terminal/zt_venv/lib/python3.9/site-packages/urwid/main_loop.py", line 385, in _run
    self.event_loop.run()
  File "/Users/kellyfoulk/Documents/code/zulip/zulip-terminal/zt_venv/lib/python3.9/site-packages/urwid/main_loop.py", line 790, in run
    self._loop()
  File "/Users/kellyfoulk/Documents/code/zulip/zulip-terminal/zt_venv/lib/python3.9/site-packages/urwid/main_loop.py", line 827, in _loop
    self._watch_files[fd]()
  File "/Users/kellyfoulk/Documents/code/zulip/zulip-terminal/zt_venv/lib/python3.9/site-packages/urwid/raw_display.py", line 416, in <lambda>
    wrapper = lambda: self.parse_input(
  File "/Users/kellyfoulk/Documents/code/zulip/zulip-terminal/zt_venv/lib/python3.9/site-packages/urwid/raw_display.py", line 515, in parse_input
    callback(processed, processed_codes)
  File "/Users/kellyfoulk/Documents/code/zulip/zulip-terminal/zt_venv/lib/python3.9/site-packages/urwid/main_loop.py", line 412, in _update
    self.process_input(keys)
  File "/Users/kellyfoulk/Documents/code/zulip/zulip-terminal/zt_venv/lib/python3.9/site-packages/urwid/main_loop.py", line 513, in process_input
    k = self._topmost_widget.keypress(self.screen_size, k)
  File "/Users/kellyfoulk/Documents/code/zulip/zulip-terminal/zulipterminal/ui.py", line 181, in keypress
    return self.controller.current_editor().keypress((size[1],), key)
  File "/Users/kellyfoulk/Documents/code/zulip/zulip-terminal/zulipterminal/ui_tools/boxes.py", line 1592, in keypress
    self.panel_view.keypress(size, 'esc')
  File "/Users/kellyfoulk/Documents/code/zulip/zulip-terminal/zulipterminal/ui_tools/views.py", line 369, in keypress
    self.log.set_focus(self.focus_index_before_search)
  File "/Users/kellyfoulk/Documents/code/zulip/zulip-terminal/zt_venv/lib/python3.9/site-packages/urwid/listbox.py", line 203, in set_focus
    self.focus = position
  File "/Users/kellyfoulk/Documents/code/zulip/zulip-terminal/zt_venv/lib/python3.9/site-packages/urwid/monitored_list.py", line 164, in _set_focus
    if index < 0 or index >= len(self):
TypeError: '<' not supported between instances of 'NoneType' and 'int'

### PR DESCRIPTION
**Pull request by Rohitth007**
This commit fixes a bug which causes Zulip to crash when searching
Streams or Topics. This happens because when a search result is
empty, log is empty and because an empty ListWalker returns None,
it leads to a Type error when setting focus.

**Commit 1**: Fixes update errors and unread count edge case.
* This is caused when ZT recieves an update event while searching and
`set_count` assumes that all widgets in the log are Buttons since
the log is updated while searching.
* This also means that message updates are only sent to a subset of
streams, topics or users. Which also leads to missed updates and
negative counts.
* Since, we only need to update the count variable in the button
object this is solved by using `btn_list` instead of `log`.

Solves part of #642 

**Commit 2**: Fix Zulip crash during search.
* This commit fixes a bug which causes Zulip to crash when searching
Streams or Topics. This happens because when a search result is
empty, log is empty and because an empty ListWalker returns None,
it leads to a Type error when setting focus.
* This is fixed by making sure the user does not enter the empty list
when the search result is empty. The boolean `empty_search` is used
to achieve this.
* This commit also improves the UI/UX when search result is empty by
displaying Text `Not Found` in place of the empty list.
The user can then correct their search and `Not Found` disappears.
* A `search_error` attribute is introducted in the pallete which
colors `Not Found` in `light red`. The `cross_mark` symbol -
`STREAM_MARKER_INVALID` is also added for effect.

Fixes #923

**Commit 3**: Name change for cross symbol for consistency.
* `STREAM_MARKER_INVALID` has been changed to `INVALID_MARKER` so
that the cross symbol can be used in multiple places not necessarily
related to streams. Eg: Search error indicator.

**Commit 4**: Change name of `topics_to_display` for consistency.
* The name of the variable `topics_to_display` has been changed to
`topic_display` to maintain consistency with `stream_display`
and `user_display`.
* This can be dropped it it feels unnecessary.

**Commit 5:** Test split for better readability.
* The tests for `PanelSearchBox` introduced when fixing bug #923 is
split into two tests for better readability.
* Both cases `test_keypress_ENTER_focus_set` and
`test_keypress_ENTER_focus_not_set` have to be checked carefully to
avoid the bug.

**Commit 6:** Erase caption when reattempting search.
* `set_content('')` is used so that when a user presses `q` for a
second search the caption becomes empty.
* This makes for a cleaner UI while searching by removing caption
`Search Results` when typing.

### CODE REVIEW

**Comment by Rohitth007** [comment]
@zulipbot add "PR needs review"

**Comment by Rohitth007** [comment]
@neiljp Does this need any more changes?

File: tests/ui_tools/test_boxes.py [hunk 1]
```diff
@@ -822,19 +824,24 @@ def test_keypress_ENTER(self, panel_search_box, widget_size,
...
+            # Switch focus to body; if have results, move to them
+            panel_search_box.panel_view.set_focus.assert_called_once_with(
+                "body")
             (panel_search_box.panel_view.body.set_focus
              .assert_called_once_with(0))
         else:
+            assert panel_search_box.caption == ''
+            (panel_search_box.panel_view.view.controller.exit_editor_mode
+             .assert_not_called())
+            panel_search_box.panel_view.set_focus.assert_not_called()
```
Comment (neiljp) [inline tests/ui_tools/test_boxes.py:844]: With these changes and large conditional branches, I'd suggest splitting the parametrize and having two separate test functions. If you do so, please have it in a separate commit first :+1:

File: zulipterminal/helper.py [hunk 2]
```diff
@@ -162,12 +162,12 @@ def _set_count_in_view(controller: Any, new_count: int,
         additionally set the current count in the model and make use of the
         same in the UI.
     """
-    stream_buttons_log = controller.view.stream_w.log
+    stream_buttons_list = controller.view.stream_w.streams_btn_list
```
Comment (neiljp) [inline zulipterminal/helper.py:165]: Could you discuss the need for these changes further? This seems like an interesting change in any case, since we update the log during search, so perhaps it belongs in a separate commit - but I'd like to hear your view on it.

File: zulipterminal/ui_tools/views.py [hunk 3]
```diff
@@ -470,6 +482,7 @@ def keypress(self, size: urwid_Size, key: str) -> Optional[str]:
 
 class UsersView(urwid.ListBox):
     def __init__(self, controller: Any, users_btn_list: List[Any]) -> None:
+        self.users_btn_list = users_btn_list
```
Comment (neiljp) [inline zulipterminal/ui_tools/views.py:485]: This is part of the set_count refactor?

**Review by neiljp — COMMENTED** [review]
@Rohitth007 This has great potential :+1: There are quite a few moving parts here, so I'd like to see them split into independent parts where they are independent, to make it clearer what each part does. The search reset seems like a separate matter entirely to the changes necessary for the UI change which resolves the bug?

File: zulipterminal/ui_tools/boxes.py [hunk 4]
```diff
@@ -1609,6 +1609,10 @@ def __init__(self, panel_view: Any, search_command: str,
         self.search_text = (
             f"Search [{', '.join(keys_for_command(search_command))}]: "
         )
+        self.search_error = urwid.AttrMap(
+            urwid.Text([' ', STREAM_MARKER_INVALID, ' Not Found']),
```
[the remark this answers is not in the document]
Comment (Rohitth007) [reply zulipterminal/ui_tools/boxes.py:1613]: Or maybe we could change its name to INVALID_OPERATION so that it can be used everywhere consistently?

File: zulipterminal/ui_tools/views.py [hunk 5]
```diff
@@ -715,6 +737,7 @@ def keypress(self, size: urwid_Size, key: str) -> Optional[str]:
         if is_command_key('SEARCH_PEOPLE', key):
             self.allow_update_user_list = False
             self.set_focus('header')
+            self.user_search.set_caption('')
```
[the remark this answers is not in the document]
Comment (Rohitth007) [reply zulipterminal/ui_tools/views.py:740]: Ok, makes sense, I'll move them :+1:

File: zulipterminal/ui_tools/views.py [hunk 6]
```diff
@@ -401,13 +408,17 @@ def update_topics(self, search_box: Any, new_text: str) -> None:
         # displaying wrong topics list.
         with self.search_lock:
             new_text = new_text.lower()
-            topics_to_display = [
+            topics_display = [
```
[the remark this answers is not in the document]
Comment (Rohitth007) [reply zulipterminal/ui_tools/views.py:411]: The rename is to keep consistency with `stream_display` and `user_display`.

File: zulipterminal/ui_tools/views.py [hunk 7]
```diff
@@ -401,13 +408,17 @@ def update_topics(self, search_box: Any, new_text: str) -> None:
...
                 topic
                 for topic in self.topics_btn_list.copy()
                 if new_text in topic.topic_name.lower()
             ]
+            self.empty_search = False if len(topics_display) else True
```
[the remark this answers is not in the document]
Comment (Rohitth007) [reply zulipterminal/ui_tools/views.py:416]: Tried ==0 but it became too long. Setting default to be `False` seems to be good in terms of space taken too. I'll make those changes.

File: zulipterminal/ui_tools/views.py (hunk 3 above)
Comment (Rohitth007) [reply zulipterminal/ui_tools/views.py:485]: Yes, this didn't have a button_list, should I move it to a separate commit?

File: zulipterminal/helper.py (hunk 2 above)
Comment (Rohitth007) [reply zulipterminal/helper.py:165]: Since, the log is updated while searching at some point we hit an attribute error because Text wrapped with AttrMap does not contain `stream_id`. Even if the log wasn't empty this would mean message updates are only sent to a subset of streams, topics or users. Since, we only need to update the count variable in the button object we can use the `btn_list` instead.

Comment (Rohitth007) [reply zulipterminal/helper.py:165]: Hopefully this doesn't conflict with the rest of the code, since a `'list'` was replaced by another `list`

File: tests/ui_tools/test_boxes.py (hunk 1 above)
Comment (Rohitth007) [reply tests/ui_tools/test_boxes.py:844]: Hmm, ok I'll split the test :+1:

File: tests/ui/test_ui_tools.py [hunk 8]
```diff
@@ -1080,13 +1080,12 @@ def test_update_user_list_editor_mode(self, mocker, right_col_view):
...
-        ('F', [], False)
-    ], ids=[
-        'user match', 'no user match',
-    ])
+    @pytest.mark.parametrize(
+        ['search_string', 'assert_list', 'match_return_value'],
+        [('U', ["USER1", "USER2"], True),
+         ('F', [], False)],
+        ids=['user match', 'no user match']
+    )
```
[the remark this answers is not in the document]
Comment (Rohitth007) [reply tests/ui/test_ui_tools.py:1088]: I move this to the other PR for refactoring `UI elements and tests` then?

File: zulipterminal/ui_tools/views.py [hunk 9]
```diff
@@ -321,17 +321,20 @@ def update_streams(self, search_box: Any, new_text: str) -> None:
             streams_display = match_stream(stream_buttons, new_text,
                                            self.view.pinned_streams)[0]
 
+            len_streams_display = len(streams_display)
```
[the remark this answers is not in the document]
Comment (Rohitth007) [reply zulipterminal/ui_tools/views.py:324]: How about `streams_display_num` ?

File: zulipterminal/helper.py (hunk 2 above)
Comment (Rohitth007) [reply zulipterminal/helper.py:165]: Actually I had a question here, though this is not related to this PR. Why do we only update count for topic only when topic view is open? Is it because there are too many topics to handle at once and instead when a user toggles into the stream, we get the latest count from the server?

File: zulipterminal/ui_tools/boxes.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Rohitth007) [reply zulipterminal/ui_tools/boxes.py:1613]: @neiljp Should I go with X or a name change as suggested above?

**Comment by Rohitth007** [comment]
@zulipbot remove "PR awaiting update"
@zulipbot add "PR needs review"

**Comment by Rohitth007** [comment]
@neiljp I have split the PR as suggested, I would like to know your views on some of the pending conversations so that they can be marked resolved.

File: tests/ui/test_ui_tools.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (neiljp) [reply tests/ui/test_ui_tools.py]: This certainly doesn't belong in the same commit, and I'd generally expect this kind of change to be with a lot of other similar changes only - which I don't expect we'll want to do unless/until we move to use something like `black`. Generally if you change code, it's reasonable to tidy it - but the point is to make each commit understandable, so extra changes don't help that.

File: zulipterminal/ui_tools/boxes.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (neiljp) [reply zulipterminal/ui_tools/boxes.py]: I wasn't necessarily meaning to change the symbol, just the text. "Not Found" suggests to me that there is one/none outcome only, but there are many potential matches here.

It would be fine to add a new symbol name, even with the same codepoint, to make this work for all the search boxes and have a more appropriate name.

File: zulipterminal/ui_tools/views.py (hunk 3 above)
Comment (neiljp) [reply zulipterminal/ui_tools/views.py:490]: If this is only necessary for the set_count refactor, then it should go with those changes.

File: zulipterminal/ui_tools/views.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (neiljp) [reply zulipterminal/ui_tools/views.py]: I think you misunderstood the second part of my point here, and I'm not sure you retested this version manually - without something like your previous version, `self.empty_search` will never get reset and ZT gets stuck with the 'Not Found' text. You need the other branch of the conditional, whether via your approach, a multiline if/else, or assignment of a boolean, eg. `self.empty_search = len(topics_display) == 0`.

File: zulipterminal/ui_tools/views.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (neiljp) [reply zulipterminal/ui_tools/views.py:414]: Consistency is good, but this code is later and reads better. If in doubt with these changes you can put them in a separate commit which often makes them easier to either drop later or squash with an adjacent commit.

File: zulipterminal/ui_tools/views.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (neiljp) [reply zulipterminal/ui_tools/views.py]: I did notice the initialization :+1:

File: zulipterminal/ui_tools/views.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (neiljp) [reply zulipterminal/ui_tools/views.py]: It's better :) There are always more options out there - this is fine.

File: tests/ui_tools/test_boxes.py (hunk 1 above)
Comment (neiljp) [reply tests/ui_tools/test_boxes.py]: You should be able to remove the parametrize completely and do some other simplification in each test now.

File: zulipterminal/helper.py (hunk 2 above)
Comment (neiljp) [reply zulipterminal/helper.py:165]: There are a few different points here:
* Normally if this just fixed a bug caused by an earlier commit in this PR, it's good practice to try and place this commit earlier as a refactor (1)
* Re the AttrMap point, I'm guessing this is related to how the `log` code works; we work around this kind of thing with the pinned-streams-divider by assigning extra attributes to a simple Divider, but...
* Most importantly, it does seem that if we're in a filtered subset then the count won't update, as you say. However, I can't reproduce this situation or a fix with this code, so we may currently have another issue which limits this fix (2)

Given the separate nature of this commit I'd suggest including it (2) but moving it earlier (1). I'm definitely interested if this corresponds to some unreads edge cases, but as I say I've not been able to make a difference with this code.

We're only iterating through a sequence, so this shouldn't make a huge difference.

Re topic view, we don't have a permanent topic view indefinitely for each stream, and the left panel switches content and loads on demand from the model, from memory.

File: tests/ui/test_ui_tools.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Rohitth007) [reply tests/ui/test_ui_tools.py]: Yes makes sense, I have already undone those changes.

File: zulipterminal/ui_tools/boxes.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Rohitth007) [reply zulipterminal/ui_tools/boxes.py]: Oh I must have skipped "No results" without noticing it. Yes No result seems more apt here, I'll change the text :+1:

File: zulipterminal/ui_tools/views.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Rohitth007) [reply zulipterminal/ui_tools/views.py]: :+1:

File: zulipterminal/ui_tools/views.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Rohitth007) [reply zulipterminal/ui_tools/views.py]: Oh oh this is a terrible mistake I made :sweat_smile: will immediately change it.
Re assignment of a boolean, I don't use this method a lot so it wasn't the first thing that came to mind. But this seems to be a much concise way of writing the same logic.

**Comment by Rohitth007** [comment]
@zulipbot remove "PR awaiting update"
@zulipbot add "PR needs review"
@neiljp I have made the suggested changes and also moved `topic_display` name change into a separate commit. If this feels unnecessary it can be dropped while merging.

File: zulipterminal/helper.py (hunk 2 above)
Comment (Rohitth007) [reply zulipterminal/helper.py:165]: Re unread edge cases, I have reproduced a particular edge case for negative counts which can be found in #642 as well as in [#zulip-terminal - Negative unread counts](https://chat.zulip.org/#narrow/stream/206-zulip-terminal/topic/Negative.20unread.20counts)

**Comment by neiljp** [comment]
@Rohitth007 Thanks for working on this! The bugfixes are great and the no-results UI is a definite improvement :+1: As per messages in the stream I've merged the unread counts bugfix and the crash bugfix/feature (plus symbol rename), with slight changes :tada:

I've left the other commits:
* the test split ended up with a lot of duplication, and the conditional looks OK to leave as-is in comparison
* the name-change isn't too important; I'd be more interested in seeing deduplication of these panel implementations in general, a little like we have for PanelSearchBox (the user-list case would need refactoring first though)
* the search improvement with the caption reset is probably part of what we want - I've brought it up in the stream under **Panel search box UI**

On that basis I'm going to close this now, and I'd suggest we build on that last commit in a new PR once we settle on a UI.

**Comment by Rohitth007** [comment]
Thanks for merging this :tada: . Having a common Panel would be brilliant, it would be easier to maintain and create new ones, but care should be taken for the user list as you said as there is no the UserView and RightColumn are mixed up. I'll look into this :+1:
